test/e2e_node/oomkiller_test.go

```diff
@@ -0,0 +1,122 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2enode
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	admissionapi "k8s.io/pod-security-admission/api"
+
+	"github.com/onsi/ginkgo/v2"
+)
+
+type testCase struct {
+	podSpec                *v1.Pod
+	oomTargetContainerName string
+}
+
+const PodOOMKilledTimeout = 2 * time.Minute
+
+var _ = SIGDescribe("OOMKiller [LinuxOnly] [NodeConformance]", func() {
+	f := framework.NewDefaultFramework("oomkiller-test")
+	f.NamespacePodSecurityEnforceLevel = admissionapi.LevelPrivileged
+
+	containerName := "oomkill-target-container"
+	oomPodSpec := getOOMTargetPod("oomkill-target-pod", containerName)
+	runOomKillerTest(f, testCase{podSpec: oomPodSpec, oomTargetContainerName: containerName})
+})
+
+func runOomKillerTest(f *framework.Framework, testCase testCase) {
+	ginkgo.Context("", func() {
+		ginkgo.BeforeEach(func() {
+			ginkgo.By("setting up the pod to be used in the test")
+			e2epod.NewPodClient(f).Create(context.TODO(), testCase.podSpec)
+		})
+
+		ginkgo.It("The containers terminated by OOM killer should have the reason set to OOMKilled", func() {
+
+			ginkgo.By("Waiting for the pod to be failed")
+			e2epod.WaitForPodTerminatedInNamespace(context.TODO(), f.ClientSet, testCase.podSpec.Name, "", f.Namespace.Name)
+
+			ginkgo.By("Fetching the latest pod status")
+			pod, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Get(context.TODO(), testCase.podSpec.Name, metav1.GetOptions{})
+			framework.ExpectNoError(err, "Failed to get the recent pod object for name: %q", pod.Name)
+
+			ginkgo.By("Verifying the OOM target container has the expected reason")
+			verifyReasonForOOMKilledContainer(pod, testCase.oomTargetContainerName)
+		})
+
+		ginkgo.AfterEach(func() {
+			ginkgo.By(fmt.Sprintf("deleting pod: %s", testCase.podSpec.Name))
+			e2epod.NewPodClient(f).DeleteSync(context.TODO(), testCase.podSpec.Name, metav1.DeleteOptions{}, framework.PodDeleteTimeout)
+		})
+	})
+}
+
+func verifyReasonForOOMKilledContainer(pod *v1.Pod, oomTargetContainerName string) {
+	container := e2epod.FindContainerStatusInPod(pod, oomTargetContainerName)
+	if container == nil {
+		framework.Failf("OOM target pod %q, container %q does not have the expected state terminated", pod.Name, container.Name)
+	}
+	if container.State.Terminated == nil {
+		framework.Failf("OOM target pod %q, container %q is not in the terminated state", pod.Name, container.Name)
+	}
+	framework.ExpectEqual(container.State.Terminated.Reason, "OOMKilled",
```

Comment:
nit: should we also check the exit code?

Reply:
Makes sense to assert on the exit code as well, pushed the change